peer/peerlist/list.go

```diff
@@ -0,0 +1,598 @@
+		pl.availableChooser.Remove(t, t.Subscriber())
+	}
+	return thunks
+}
+
+// removeAllUnavailablePeers will clear the unavailablePeers list and
+// return all the Peers in the list in a slice
+// Must be run in a mutex.Lock()
+func (pl *List) removeAllUnavailablePeers(toRemove map[string]*peerThunk) []*peerThunk {
+	thunks := make([]*peerThunk, 0, len(toRemove))
+	for id, t := range toRemove {
+		thunks = append(thunks, t)
+		delete(toRemove, id)
+	}
+	return thunks
+}
+
+// releaseAll will iterate through a list of peers and call release
+// on the transport
+func (pl *List) releaseAll(peers []*peerThunk) []error {
+	var errs []error
+	for _, t := range peers {
+		if err := pl.transport.ReleasePeer(t.peer, t); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+// removePeerIdentifier will go remove references to the peer identifier and release
+// it from the transport
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifier(pid peer.Identifier) error {
+	t, err := pl.removePeerIdentifierReferences(pid)
+	if err != nil {
+		// The peer has already been removed
+		return err
+	}
+
+	return pl.transport.ReleasePeer(pid, t)
+}
+
+// removePeerIdentifierReferences will search through the Available and Unavailable Peers
+// for the PeerID and remove it
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifierReferences(pid peer.Identifier) (*peerThunk, error) {
+	if t := pl.availablePeers[pid.Identifier()]; t != nil {
+		return t, pl.removeFromAvailablePeers(t)
+	}
+
+	if t, ok := pl.unavailablePeers[pid.Identifier()]; ok && t != nil {
+		pl.removeFromUnavailablePeers(t)
+		return t, nil
+	}
+
+	return nil, peer.ErrPeerRemoveNotInList(pid.Identifier())
+}
+
+// removeFromAvailablePeers remove a peer from the Available Peers list the
+// Peer should already be validated as non-nil and in the Available list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromAvailablePeers(t *peerThunk) error {
+	delete(pl.availablePeers, t.peer.Identifier())
+	pl.availableChooser.Remove(t, t.Subscriber())
+	t.SetSubscriber(nil)
+	return nil
+}
+
+// removeFromUnavailablePeers remove a peer from the Unavailable Peers list the
+// Peer should already be validated as non-nil and in the Unavailable list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromUnavailablePeers(t *peerThunk) {
+	delete(pl.unavailablePeers, t.peer.Identifier())
+}
+
+// Choose selects the next available peer in the peer list
+func (pl *List) Choose(ctx context.Context, req *transport.Request) (peer.Peer, func(error), error) {
+	if err := pl.once.WaitUntilRunning(ctx); err != nil {
+		return nil, nil, pl.newNotRunningError(err)
+	}
+
+	for {
+		p, err := pl.choose(ctx, req)
+		if p != nil {
+			t := p.(*peerThunk)
+			pl.notifyPeerAvailable()
+			t.onStart()
+			return t.peer, t.boundOnFinish, nil
+		}
+		if !yarpcerrors.IsUnavailable(err) {
+			return nil, nil, err
+		}
+		if err := pl.waitForPeerAddedEvent(ctx); err != nil {
```

Comment:
So this function is only applicable if there are no "available" peers.  If we hit this function because the availableChooser returned nil/error but where we do have available peers we will block the request until the context times out, or a new peer is made available (which is not guaranteed).

If an available chooser wants to wait for availability it will need to do that itself.

We have the option of only blocking here if there are no available peers, but that provides an opportunity here for an infinite loop if the availableChooser gets in a state where it's continuously returning errors... :(

It might make sense to have each availableChooser perform the blocking themselves (this includes the whole notify syntax etc).  We can take the core "notify" functionality and put it into a package that can get embedded so that people don't need to play around with all the channels.  We'd probably need to make this a separate API entry point though, since if we put the "block" in the choose method we're likely to have a deadlock.

Reply:
I’m just going to write into the contract that ListImplementation must return a peer if any are available and if you need something different, you’ll need to write your own peer chooser.